// telnyx/resources/addresses/addresses.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Addresses = void 0;
const tslib_1 = require("../../internal/tslib.js");
const resource_1 = require("../../core/resource.js");
const ActionsAPI = tslib_1.__importStar(require("./actions.js"));
const actions_1 = require("./actions.js");
const pagination_1 = require("../../core/pagination.js");
const path_1 = require("../../internal/utils/path.js");
class Addresses extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.actions = new ActionsAPI.Actions(this._client);
    }
    /**
     * Creates an address.
     *
     * @example
     * ```ts
     * const address = await client.addresses.create({
     *   business_name: "Toy-O'Kon",
     *   country_code: 'US',
     *   first_name: 'Alfred',
     *   last_name: 'Foster',
     *   locality: 'Austin',
     *   street_address: '600 Congress Avenue',
     * });
     * ```
     */
    create(body, options) {
        return this._client.post('/addresses', { body, ...options });
    }
    /**
     * Retrieves the details of an existing address.
     *
     * @example
     * ```ts
     * const address = await client.addresses.retrieve('id');
     * ```
     */
    retrieve(id, options) {
        return this._client.get((0, path_1.path) `/addresses/${id}`, options);
    }
    /**
     * Returns a list of your addresses.
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const address of client.addresses.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/addresses', (pagination_1.DefaultFlatPagination), { query, ...options });
    }
    /**
     * Deletes an existing address.
     *
     * @example
     * ```ts
     * const address = await client.addresses.delete('id');
     * ```
     */
    delete(id, options) {
        return this._client.delete((0, path_1.path) `/addresses/${id}`, options);
    }
}
exports.Addresses = Addresses;
Addresses.Actions = actions_1.Actions;
