// telnyx/resources/conferences/actions.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../core/resource.mjs";
import { path } from "../../internal/utils/path.mjs";
export class Actions extends APIResource {
    /**
     * Update conference participant supervisor_role
     *
     * @example
     * ```ts
     * const action = await client.conferences.actions.update(
     *   'id',
     *   {
     *     call_control_id:
     *       'v3:MdI91X4lWFEs7IgbBEOT9M4AigoY08M0WWZFISt1Yw2axZ_IiE4pqg',
     *     supervisor_role: 'whisper',
     *   },
     * );
     * ```
     */
    update(id, body, options) {
        return this._client.post(path `/conferences/${id}/actions/update`, { body, ...options });
    }
    /**
     * Hold a list of participants in a conference call
     *
     * @example
     * ```ts
     * const response = await client.conferences.actions.hold(
     *   'id',
     * );
     * ```
     */
    hold(id, body, options) {
        return this._client.post(path `/conferences/${id}/actions/hold`, { body, ...options });
    }
    /**
     * Join an existing call leg to a conference. Issue the Join Conference command
     * with the conference ID in the path and the `call_control_id` of the leg you wish
     * to join to the conference as an attribute. The conference can have up to a
     * certain amount of active participants, as set by the `max_participants`
     * parameter in conference creation request.
     *
     * **Expected Webhooks:**
     *
     * - `conference.participant.joined`
     * - `conference.participant.left`
     *
     * @example
     * ```ts
     * const response = await client.conferences.actions.join(
     *   'id',
     *   {
     *     call_control_id:
     *       'v3:MdI91X4lWFEs7IgbBEOT9M4AigoY08M0WWZFISt1Yw2axZ_IiE4pqg',
     *   },
     * );
     * ```
     */
    join(id, body, options) {
        return this._client.post(path `/conferences/${id}/actions/join`, { body, ...options });
    }
    /**
     * Removes a call leg from a conference and moves it back to parked state.
     *
     * **Expected Webhooks:**
     *
     * - `conference.participant.left`
     *
     * @example
     * ```ts
     * const response = await client.conferences.actions.leave(
     *   'id',
     *   {
     *     call_control_id: 'c46e06d7-b78f-4b13-96b6-c576af9640ff',
     *   },
     * );
     * ```
     */
    leave(id, body, options) {
        return this._client.post(path `/conferences/${id}/actions/leave`, { body, ...options });
    }
    /**
     * Mute a list of participants in a conference call
     *
     * @example
     * ```ts
     * const response = await client.conferences.actions.mute(
     *   'id',
     * );
     * ```
     */
    mute(id, body, options) {
        return this._client.post(path `/conferences/${id}/actions/mute`, { body, ...options });
    }
    /**
     * Play audio to all or some participants on a conference call.
     *
     * @example
     * ```ts
     * const response = await client.conferences.actions.play(
     *   'id',
     * );
     * ```
     */
    play(id, body, options) {
        return this._client.post(path `/conferences/${id}/actions/play`, { body, ...options });
    }
    /**
     * Pause conference recording.
     *
     * @example
     * ```ts
     * const response =
     *   await client.conferences.actions.recordPause('id');
     * ```
     */
    recordPause(id, body, options) {
        return this._client.post(path `/conferences/${id}/actions/record_pause`, { body, ...options });
    }
    /**
     * Resume conference recording.
     *
     * @example
     * ```ts
     * const response =
     *   await client.conferences.actions.recordResume('id');
     * ```
     */
    recordResume(id, body, options) {
        return this._client.post(path `/conferences/${id}/actions/record_resume`, { body, ...options });
    }
    /**
     * Start recording the conference. Recording will stop on conference end, or via
     * the Stop Recording command.
     *
     * **Expected Webhooks:**
     *
     * - `conference.recording.saved`
     *
     * @example
     * ```ts
     * const response =
     *   await client.conferences.actions.recordStart('id', {
     *     format: 'wav',
     *   });
     * ```
     */
    recordStart(id, body, options) {
        return this._client.post(path `/conferences/${id}/actions/record_start`, { body, ...options });
    }
    /**
     * Stop recording the conference.
     *
     * **Expected Webhooks:**
     *
     * - `conference.recording.saved`
     *
     * @example
     * ```ts
     * const response =
     *   await client.conferences.actions.recordStop('id');
     * ```
     */
    recordStop(id, body, options) {
        return this._client.post(path `/conferences/${id}/actions/record_stop`, { body, ...options });
    }
    /**
     * Convert text to speech and play it to all or some participants.
     *
     * @example
     * ```ts
     * const response = await client.conferences.actions.speak(
     *   'id',
     *   { payload: 'Say this to participants', voice: 'female' },
     * );
     * ```
     */
    speak(id, body, options) {
        return this._client.post(path `/conferences/${id}/actions/speak`, { body, ...options });
    }
    /**
     * Stop audio being played to all or some participants on a conference call.
     *
     * @example
     * ```ts
     * const response = await client.conferences.actions.stop(
     *   'id',
     * );
     * ```
     */
    stop(id, body, options) {
        return this._client.post(path `/conferences/${id}/actions/stop`, { body, ...options });
    }
    /**
     * Unhold a list of participants in a conference call
     *
     * @example
     * ```ts
     * const response = await client.conferences.actions.unhold(
     *   'id',
     *   {
     *     call_control_ids: [
     *       'v3:MdI91X4lWFEs7IgbBEOT9M4AigoY08M0WWZFISt1Yw2axZ_IiE4pqg',
     *     ],
     *   },
     * );
     * ```
     */
    unhold(id, body, options) {
        return this._client.post(path `/conferences/${id}/actions/unhold`, { body, ...options });
    }
    /**
     * Unmute a list of participants in a conference call
     *
     * @example
     * ```ts
     * const response = await client.conferences.actions.unmute(
     *   'id',
     * );
     * ```
     */
    unmute(id, body, options) {
        return this._client.post(path `/conferences/${id}/actions/unmute`, { body, ...options });
    }
}
