// telnyx/resources/ai/ai.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../core/resource.mjs";
import * as AudioAPI from "./audio.mjs";
import { Audio } from "./audio.mjs";
import * as ChatAPI from "./chat.mjs";
import { Chat } from "./chat.mjs";
import * as ClustersAPI from "./clusters.mjs";
import { Clusters, } from "./clusters.mjs";
import * as McpServersAPI from "./mcp-servers.mjs";
import { McpServers, } from "./mcp-servers.mjs";
import * as AssistantsAPI from "./assistants/assistants.mjs";
import { Assistants, } from "./assistants/assistants.mjs";
import * as ConversationsAPI from "./conversations/conversations.mjs";
import { Conversations, } from "./conversations/conversations.mjs";
import * as EmbeddingsAPI from "./embeddings/embeddings.mjs";
import { Embeddings, } from "./embeddings/embeddings.mjs";
import * as FineTuningAPI from "./fine-tuning/fine-tuning.mjs";
import { FineTuning } from "./fine-tuning/fine-tuning.mjs";
import * as IntegrationsAPI from "./integrations/integrations.mjs";
import { Integrations, } from "./integrations/integrations.mjs";
export class AI extends APIResource {
    constructor() {
        super(...arguments);
        this.assistants = new AssistantsAPI.Assistants(this._client);
        this.audio = new AudioAPI.Audio(this._client);
        this.chat = new ChatAPI.Chat(this._client);
        this.clusters = new ClustersAPI.Clusters(this._client);
        this.conversations = new ConversationsAPI.Conversations(this._client);
        this.embeddings = new EmbeddingsAPI.Embeddings(this._client);
        this.fineTuning = new FineTuningAPI.FineTuning(this._client);
        this.integrations = new IntegrationsAPI.Integrations(this._client);
        this.mcpServers = new McpServersAPI.McpServers(this._client);
    }
    /**
     * This endpoint returns a list of Open Source and OpenAI models that are available
     * for use. <br /><br /> **Note**: Model `id`'s will be in the form
     * `{source}/{model_name}`. For example `openai/gpt-4` or
     * `mistralai/Mistral-7B-Instruct-v0.1` consistent with HuggingFace naming
     * conventions.
     *
     * @example
     * ```ts
     * const response = await client.ai.retrieveModels();
     * ```
     */
    retrieveModels(options) {
        return this._client.get('/ai/models', options);
    }
    /**
     * Generate a summary of a file's contents.
     *
     * Supports the following text formats:
     *
     * - PDF, HTML, txt, json, csv
     *
     * Supports the following media formats (billed for both the transcription and
     * summary):
     *
     * - flac, mp3, mp4, mpeg, mpga, m4a, ogg, wav, or webm
     * - Up to 100 MB
     *
     * @example
     * ```ts
     * const response = await client.ai.summarize({
     *   bucket: 'bucket',
     *   filename: 'filename',
     * });
     * ```
     */
    summarize(body, options) {
        return this._client.post('/ai/summarize', { body, ...options });
    }
}
AI.Assistants = Assistants;
AI.Audio = Audio;
AI.Chat = Chat;
AI.Clusters = Clusters;
AI.Conversations = Conversations;
AI.Embeddings = Embeddings;
AI.FineTuning = FineTuning;
AI.Integrations = Integrations;
AI.McpServers = McpServers;
